src/sema/eval.rs

```diff
@@ -196,3 +202,283 @@ pub fn eval_const_rational(
+                    ty,
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, bool) => {
+            let left = eval_constants_in_expression(*left)?;
+            let right = eval_constants_in_expression(*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right.clone(), Type::Uint(64), *right_loc)
+                {
+                    Err(diagnostic)
+                } else {
+                    Ok(Expression::NumberLiteral(
+                        loc,
+                        ty,
+                        left.shr(right.to_u64().unwrap()),
+                    ))
+                }
+            } else {
+                Ok(Expression::ShiftRight(
+                    loc,
+                    ty,
+                    Box::new(left),
+                    Box::new(right),
+                    bool,
+                ))
+            }
+        }
+        Expression::Builtin(.., ref args) => {
+            for args_iter in args {
+                let expression = eval_constants_in_expression(args_iter.clone())?;
+                if let Expression::NumberLiteral(loc, ty, res) = expression {
+                    if let Some(diagnostic) = overflow_check(res, ty, loc) {
+                        return Err(diagnostic);
+                    }
+                }
+            }
+
+            Ok(expr)
+        }
+
+        _ => Ok(expr),
+    }
+}
+
+// Function that takes a BigInt and an expected type. If the number of bits in the type required to represent the BigInt is not suffiecient, it will return a diagnostic.
+fn overflow_check(result: BigInt, ty: Type, loc: Loc) -> Option<Diagnostic> {
+    if let Type::Uint(bits) = ty {
+        // If the result sign is minus, throw an error.
+        if let Sign::Minus = result.sign() {
+            return Some(Diagnostic::error(
+                loc,
+            format!( "negative value {} does not fit into type uint{}. Cannot implicitly convert signed literal to unsigned type.",result,ty.get_type_size()),
+            ));
+        }
+
+        // If bits of the result is more than bits of the type, throw and error.
+        if result.bits() > bits as u64 {
+            return Some(Diagnostic::error(
+                loc,
+                format!(
+                    "value {} does not fit into type uint{}.",
+                    result,
+                    ty.get_type_size(),
+                ),
+            ));
+        }
+    }
+
+    if let Type::Int(bits) = ty {
+        // If number of bits is more than what the type can hold. BigInt.bits() is not used here since it disregards the sign.
+        if result.to_signed_bytes_be().len() * 8 > (bits as usize) {
```

Comment:
If the sign is negative, you'll need `result.bits() + 1` to store the number. By doing `result.to_signed_bytes_be()` you are copying the inner data data structure that holds the number. That is not necessary for this case.

Reply:
This will fail for the lower boundary: `int8 sesa = -128` won't be allowed
I have to add a special condition for this If I use `res.bits() + 1` so I thought this is more neat

Reply:
This will not work for the boundaries.
i.e -128 will not fit into int8 using bits() + 1

Reply:
This should be permitted:
```rust
int8 foo = -128;
```
is it?